doc/docs/Parallel_Meep.md

```diff
@@ -169,3 +169,136 @@ See also [FAQ/Should I expect linear speedup from the parallel Meep](FAQ.md#shou
+
+<center>
+![](images/adaptive_chunk_layout.gif)
+</center>
+
+### Chunk balancer interface
+
+The chunk balancer interface provides four main methods:
+- `make_initial_chunk_layout()` generates the initial chunk layout for the first iteration of a simulation. By default, `None` is returned, indicating Meep should use its default chunk partitioning strategy.
+- `should_adjust_chunk_layout()` decides whether the current layout is sufficiently poorly balanced to justify the up-front cost of reallocating the field arrays when changing chunk layouts.
+- `compute_new_chunk_layout()` looks at the current chunk layout and per-process timing data to compute a new chunk layout which has chunk sizes adjusted to improve the load balance across compute nodes.
+- `adjust_chunk_layout()` is syntactic sugar which will compute a new chunk layout, apply it to the simulation object, reset the simulation, and re-initialize the simulation.
+
+```py
+class AbstractChunkBalancer(abc.ABC):
+  """Chunk balancer for dynamically load-balanced chunk layouts."""
+
+  def make_initial_chunk_layout(self, sim) -> mp.BinaryPartition:
+    """Generates an initial chunk layout for simulation."""
+
+  def should_adjust_chunk_layout(self, sim) -> bool:
+    """Is current layout imbalanced enough to justify rebuilding sim?"""
+
+  @abc.abstractmethod
+  def compute_new_chunk_layout(
+    self,
+    timing_measurements: MeepTimingMeasurements,
+    old_chunk_layout: mp.BinaryPartition,
+    chunk_volumes: Tuple[mp.grid_volume],
+    chunk_owners: np.ndarray) -> mp.BinaryPartition:
+    """Rebalance chunks to equalize simulation time for each node."""
+
+  def adjust_chunk_layout(self, sim, **kwargs) -> None:
+    """Computes a new chunk layout and applies it to sim."""
+```
+
+### Usage
+
+Using the chunk balancer is very straightforward, and it can typically be integrated into existing Meep simulations with only a few lines of code. Here is a simple example:
+
+```py
+from meep.chunk_balancer import ChunkBalancer
+
+chunk_balancer = ChunkBalancer()
+
+# Compute an initial chunk layout
+initial_chunk_layout = chunk_balancer.make_initial_chunk_layout()
+
+sim = mp.Simulation(..., chunk_layout=initial_chunk_layout)
+sim.init_sim()
+
+for iteration in range(epochs):
+  sim.run(...)
+  # Adjust the chunk layout for the next iteration if needed
+  chunk_balancer.adjust_chunk_layout(sim, sensitivity=0.4)
+```
+
+Chunks can also be rebalanced between runs of a program by dumping and loading the chunk layout from a pickled object. For example:
+
+```py
+import meep as mp
+from meep.chunk_balancer import ChunkBalancer
+from meep.timing_measurements import MeepTimingMeasurements
+import pickle
+
+# Fetch chunk layout from a previous run
+with open("path/to/chunk_layout.pkl", "rb") as f:
+    chunk_layout = pickle.load(f)
```

Comment:
Should set `chunk_layout = None` if the file doesn't exist.

Reply:
Done